src/binary_format.cc

```diff
@@ -32,144 +35,165 @@ FactorGraphDescriptor read_meta(const std::string &meta_file) {
   return meta;
 }
 
-void FactorGraph::load_weights(const std::string &filename) {
-  std::ifstream file(filename, std::ios::binary);
-  size_t count = 0;
-  while (file && file.peek() != EOF) {
-    // read fields
-    size_t wid;
-    uint8_t isfixed;
-    double initial_value;
-    read_be_or_die(file, wid);
-    read_be_or_die(file, isfixed);
-    read_be_or_die(file, initial_value);
-    // load into factor graph
-    weights[wid] = Weight(wid, initial_value, isfixed);
-    ++count;
-  }
-  size.num_weights += count;
+#define PARALLEL_LOAD(...)                                              \
+  std::atomic<size_t> counter, counter2;                                \
+  counter = counter2 = 0;                                               \
+  std::mutex lock;                                                      \
+  std::vector<std::thread> threads;                                     \
+  for (auto &filename : filenames) {                                    \
+    threads.push_back(std::thread(                                      \
+        [&counter, &counter2, &lock, this, filename] { __VA_ARGS__ })); \
+  }                                                                     \
+  for (auto &t : threads) t.join();                                     \
+  threads.clear();
```

Comment:
It seems the macro rather adds more complexity than removes from the rest of the code that use it.  Besides, it's generally recommended against passing statements as arguments. If you don't mind I want to amend things a bit.